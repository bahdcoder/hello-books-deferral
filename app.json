{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
