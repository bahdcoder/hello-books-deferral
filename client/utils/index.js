export const getDateStr = str => {
    const end = str.indexOf('T');
    const rawDate = str.slice(0, end);
    const firstHyphen = rawDate.indexOf('-');
    const lastHyphen = rawDate.lastIndexOf('-');
    const year = rawDate.slice(0, firstHyphen);
    const month = rawDate.slice(firstHyphen + 1, lastHyphen);
    const day = rawDate.slice(lastHyphen + 1);
    const monthStrArr = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
    ];
    const monthStr = monthStrArr[Number(month) - 1];
    return `${monthStr} ${day} ${year}`;
};
