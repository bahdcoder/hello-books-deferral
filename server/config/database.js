export default () => ({
    development: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    test: {
        client: process.env.TEST_DATABASE_CLIENT,
        connection: process.env.TEST_DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    review: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.BACKUP_DATABASE,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    staging: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    },

    production: {
        client: process.env.DATABASE_CLIENT,
        connection: process.env.DATABASE_URL,
        pool: {
            min: 2,
            max: 10
        },
        migrations: {
            tableName: 'knex_migrations',
            directory: './server/migrations'
        }
    }
});
