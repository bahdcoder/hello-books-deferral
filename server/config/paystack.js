export default () => ({
    secret: process.env.PAYSTACK_SECRET_KEY,
    public: process.env.PAYSTACK_PUBLIC_KEY
});
