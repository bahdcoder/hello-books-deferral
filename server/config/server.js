export default () => ({
  port: process.env.PORT || 5000,
  env: process.env.NODE_ENV || "development",
  url: process.env.APP_URL || "http://localhost:5000"
});
