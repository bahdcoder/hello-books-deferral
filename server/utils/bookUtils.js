const { validations } = require("indicative");
const { Vanilla } = require("indicative/builds/formatters");
const Validator = require("indicative/builds/validator");
const Book = require("../models/Book");

/* custom user friendly error messages */
const messages = {
  required: "{{ field }} is required to create a book",
  string: "{{ field }} is not a string",
  number: "{{ field }} is not a number",
  different: "{{ field }} must be unique"
};

// Trim white spaces
const sanitizeRules = {
  title: "trim",
  coverType: "trim",
  description: "trim",
  isbn: "trim",
  publisher: "trim",
  year: "trim",
  copiesAvailable: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.different = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  const row = await Book.query().where(column, value);
  if (row[0]) throw message;
};

const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */

module.exports = { validatorInstance, messages, sanitizeRules };
