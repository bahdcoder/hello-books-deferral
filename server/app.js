import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
