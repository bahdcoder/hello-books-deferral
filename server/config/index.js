import dotenv from "dotenv";
import auth from "@config/auth";
import server from "@config/server";
import database from "@config/database";

dotenv.config();

export default {
  auth: auth(),
  server: server(),
  database: database()
};
