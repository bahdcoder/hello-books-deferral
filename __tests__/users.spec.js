import supertest from 'supertest';
import { app, databaseConnection } from '@server/app';
import { getUser, createUser, superAdminUser } from '@tests/utils/helpers';
import jwt from 'jsonwebtoken';
import config from '@config';

const server = () => supertest(app);

describe('USER API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection.migrate.latest();
    });
    it('super admin should be able to get users', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const { status } = await server()
            .get('/api/v1/users?page=0')
            .set('x-access-token', token);

        expect(status).toBe(200);
    });

    it('Super Admin can update a user field', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const norm_user = getUser();
        const { id } = await createUser(norm_user);

        const { status, body } = await server()
            .patch(`/api/v1/users/update/${id}`)
            .set('x-access-token', token)
            .send({
                ...getUser()
            });

        expect(status).toBe(200);
        expect(Object.keys(body)).toMatchSnapshot();
    });

    it('should throw an error if update fields are missing', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const norm_user = getUser();
        const { id } = await createUser(norm_user);

        const { status, body } = await server()
            .patch(`/api/v1/users/update/${id}`)
            .set('x-access-token', token)
            .send({});

        expect(status).toBe(422);
        expect(Object.keys(body)).toMatchSnapshot();
    });

    it('super admin can change access level of normal users', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const norm_user = getUser();
        const { id } = await createUser(norm_user);

        const { status, body } = await server()
            .patch(`/api/v1/users/change-access-level/${id}`)
            .set('x-access-token', token)
            .send({
                role: 'admin'
            });

        expect(status).toBe(200);
        expect(Object.keys(body)).toMatchSnapshot();
    });

    it('super admin can delete a user', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const norm_user = getUser();
        const { id } = await createUser(norm_user);

        const { status, body } = await server()
            .delete(`/api/v1/users/${id}`)
            .set('x-access-token', token);

        expect(status).toBe(200);
        expect(Object.keys(body)).toMatchSnapshot();
    });

    it('should return an error if user does not exist', async () => {
        const user = getUser();
        const admin = await superAdminUser(user);

        const token = jwt.sign(
            { id: admin.id, email: admin.email },
            config.auth.secret
        );

        const { status, body } = await server()
            .delete(`/api/v1/users/474729238`)
            .set('x-access-token', token);

        expect(status).toBe(400);
        expect(Object.keys(body)).toMatchSnapshot();
    });
});
