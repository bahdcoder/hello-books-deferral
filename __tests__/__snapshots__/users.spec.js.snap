// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USER API ENDPOINTS Super Admin can update a user field 1`] = `
Array [
  "status",
  "data",
]
`;

exports[`USER API ENDPOINTS should return an error if user does not exist 1`] = `
Array [
  "status",
  "data",
]
`;

exports[`USER API ENDPOINTS should throw an error if update fields are missing 1`] = `
Array [
  "status",
  "code",
  "message",
]
`;

exports[`USER API ENDPOINTS super admin can change access level of normal users 1`] = `
Array [
  "status",
  "data",
]
`;

exports[`USER API ENDPOINTS super admin can change access level of normal users 2`] = `
Array [
  "status",
  "data",
]
`;

exports[`USER API ENDPOINTS super admin can delete a user 1`] = `
Array [
  "status",
  "data",
]
`;
